src/ship/GunManager.h

```diff
@@ -0,0 +1,222 @@
+ * - A higher-level input manager sends fire commands for each group
+ * - Weapon mounts are added or removed as needed by equipment attached to the ship
+ * - Finally, when the ship is destroyed or its type changed, the GunManager is destroyed
+ */
+class GunManager : public LuaWrappable {
+public:
+	static constexpr size_t NUM_WEAPON_INDICES = 256;
+	using WeaponIndexSet = std::bitset<NUM_WEAPON_INDICES>;
+
+	struct WeaponMount;
+
+	enum ProjectileType : uint8_t { // <enum scope='GunManager' name='ProjectileType' prefix='PROJECTILE_' public>
+		PROJECTILE_BALLISTIC, // Impact weapons: pulsecannons, railguns, etc.
+		PROJECTILE_BEAM,      // Instantaneous-hit weapons: lasers
+	};
+
+	struct ProjectileDef {
+		float lifespan = 5.0;
+		float speed = 1000.0;
+
+		// Damage
+		float impactDamage = 0.0; // How much damage is dealt when this projectile hits a target
+		float explosiveDamage = 0.0; // When this projectile detonates (impact or proximity fuse) how much damage is dealt in the radius?
+		float explosionRadius = 0.0; // How large of an explosion radius does this projectile have?
+		float explosionFalloff = 1.0; // How does the damage fall off from the center to the edge of the explosion? 2.0 = quadratic, 1.0 = linear, 0.0 = no falloff, use "impact area" approximation instead
+		// TODO: damage type
+
+		// Feature flags
+		uint8_t isMining : 1; // Does this projectile deal "mining" damage? TODO: better damage type system
+		uint8_t hasProxyFuse : 1; // Does this projectile have a proximity fuse?
+		uint8_t proxyIFF : 1; // Is the proximity fuse capable of reading IFF data?
+
+		// Fusing
+		float proxyFuseRadius = 0.0; // This projectile will detonate when it detects a body within this range
+		float proxyFuseArmTime = 0.5; // How long after firing before the projectile fuse is armed?
+
+		// Visual settings
+		float length = 0;
+		float width = 0;
+		Color color;
+	};
+
+	// Information about a specific "weapon type"
+	// TODO: create one of these in Lua per weapon definition and reference them from each mounted gun
+	// TODO: create a separate projectile definition and reference it
+	struct WeaponData {
+		float firingRPM = 240;	// number of shots per minute (60s / time between shots)
+		float firingHeat = 4.5; // amount of thermal energy(kJ) emitted into the system per shot
+
+		//TODO: integrate this with a whole-ship cooling system
+		float coolingPerSecond = 14.2; // nominal amount of thermal energy removed per second (kW)
+		float overheatThreshold = 280; // total amount of thermal energy(kJ) the gun can store while functioning
+
+		std::string modelPath; // model to render this weapon with
+
+		ProjectileData projectile; // deprecated, to replace with RefCountedPtr<ProjectileDef>
+		ProjectileType projectileType = PROJECTILE_BALLISTIC;
+		uint8_t numBarrels = 1;		 // total number of barrels on the model
+		bool staggerBarrels = false; // should we fire one barrel after another, or both at the same time?
```

Comment:
What are all these numbers for (as in, the comments say what the variable is, but these numbers look like the concrete implementation of a specific weapon...?)

Reply:
These are "sane defaults" that were added as an example during prototyping - I can amend these to be initialized with valid but nonspecific values instead. It is intended that all fields of a `WeaponData` object are filled by external code before it is used by the GunManager, so these defaults are not particularly useful.